core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -812,40 +816,48 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
         }
     }
 
-    private static final int[] EMPTY_DOC_IDS = new int[0];
-
     /**
      * Shortcut ids to load, we load only "from" and up to "size". The phase controller
      * handles this as well since the result is always size * shards for Q_A_F
      */
     private void shortcutDocIdsToLoad(SearchContext context) {
+        final int[] searchDocIdsToLoad;
         if (context.request().scroll() != null) {
             TopDocs topDocs = context.queryResult().topDocs();
-            int[] docIdsToLoad = new int[topDocs.scoreDocs.length];
+            searchDocIdsToLoad = new int[topDocs.scoreDocs.length];
             for (int i = 0; i < topDocs.scoreDocs.length; i++) {
-                docIdsToLoad[i] = topDocs.scoreDocs[i].doc;
+                searchDocIdsToLoad[i] = topDocs.scoreDocs[i].doc;
             }
-            context.docIdsToLoad(docIdsToLoad, 0, docIdsToLoad.length);
         } else {
             TopDocs topDocs = context.queryResult().topDocs();
             if (topDocs.scoreDocs.length < context.from()) {
                 // no more docs...
-                context.docIdsToLoad(EMPTY_DOC_IDS, 0, 0);
-                return;
-            }
-            int totalSize = context.from() + context.size();
-            int[] docIdsToLoad = new int[Math.min(topDocs.scoreDocs.length - context.from(), context.size())];
-            int counter = 0;
-            for (int i = context.from(); i < totalSize; i++) {
-                if (i < topDocs.scoreDocs.length) {
-                    docIdsToLoad[counter] = topDocs.scoreDocs[i].doc;
-                } else {
-                    break;
+                searchDocIdsToLoad = new int[0];
+            } else {
+                int totalSize = context.from() + context.size();
+                searchDocIdsToLoad = new int[Math.min(topDocs.scoreDocs.length - context.from(), context.size())];
+                for (int i = context.from(); i < Math.min(totalSize, topDocs.scoreDocs.length); i++) {
+                    searchDocIdsToLoad[i - context.from()] = topDocs.scoreDocs[i].doc;
+                }
+            }
+        }
+        int[] docIdsToLoad = searchDocIdsToLoad;
+        final Suggest suggest = context.queryResult().suggest();
+        if (suggest != null && suggest.hasScoreDocs()) {
```

Comment:
I think what would make this simpler is if we had an API on `List<CompletionSuggestion> suggestions = suggest.filter(CompletionSuggestion.class)` that we can use to iterate over twice:
```
int numSuggestDocs = 0;
for (CompletionSuggestion completionSuggestion : suggestions) {
  numSuggestDocs += completionSuggestion.getScoreDocs().length
}

// now fill the docIDs from the search and use the `numSuggestDocs` to extend the size of the array.
```
makes sense?